Fixing a few issues with tracks events in gateway suggestions

joelclimbsthings opened the pull request:
Fixes #7303

As per the original issue, some of the tracks events are not being fired quite right. Here is an audit of the events:

## Initial Audit

**wcadmin_task_view**
Currently firing as expected.

> Is it possible to add a new prop showing all the payment methods options the user is seeing when they visit the payments task?

This _would_ be possible, but this event is currently decoupled from the payment gateways suggestions feature entirely, and the data in terms of visible gateways isn't available when fired. Perhaps we could add another event that only fires on loading the payment task?

**wcadmin_tasklist_payment_setup**
This event wasn't firing at all due to an earlier typo 🤭 . This is fixed, although I want to note that the gateway IDs included in the payload _only_ include those that appeared in the same list (Enabled/Additional), and never includes `woocommerce_payments` since it always appears outside of the lists and in it's own card. We could amend this to include all visible gateways if preferred, but I haven't made this change yet.

**wcadmin_tasklist_payment_install_method**
Firing as expected, but only when plugins are installed. This does _not_ include BACS and COD since those don't require installation. This seems like the correct behavior semantically; do we want to change this?

**wcadmin_payments_task_stepper_view**
This seems to be firing as expected. Only when Setup is required for a gateway, and it's being displayed.

**wcadmin_tasklist_payment_connect_start**
This was being fired whenever a connect form was submitted, although by the description it sounds like it's expected to fire when a _Connect_ button directs the user away from the site (Stripe/etc), so I've moved it to execute in this instance only. Let me know if this is not correct.

**wcadmin_tasklist_payment_connect_method**
This occurs for all methods when they're marked as configured within the payments task, as expected I believe.

## Current event flows
_Note: I'll updated this as changes are made in this PR._

**wcadmin_tasklist_payments_options**
Fires when payment task loads, and includes all visible gateway IDs in payload.

**wcadmin_tasklist_payment_setup**
Fires only when "Setup" or "Finish" setup buttons are clicked, so only for gateways that include Setup in the task. Currently includes all IDs of gateways _that appeared in the same list_ (Additional/Enabled/etc), as well as the ID of the selected gateway.

**wcadmin_payments_task_stepper_view**
This fires whenever the setup Stepper is displayed for a gateway. Includes the ID of the gateway.

**wcadmin_tasklist_payment_install_method**
Only fires when a plugin is installed for a gateway as part of the setup. This includes a list of slugs for all plugins installed.

**wcadmin_tasklist_payment_connect_start**
Currently (as of this PR) is firing when a gateway provides a connection URL. This will result in us displaying a "Connect" button during setup, which will fire this event when clicked (and direct the user to the connection URL). Includes ID of gateway.

**wcadmin_tasklist_payment_connect_method**
This is fired whenever a gateway is marked as configured, which _should_ occur for all gateways once they're connected/enabled. Includes ID of gateway.

### Detailed test instructions:

-   Checkout branch
-   Enter this in the console -> `localStorage.setItem( 'debug', 'wc-admin:*' );`
- Go through the basic flows described above for each event and ensure it's fired as expected.

pmcpinto commented:
@joelclimbsthings thanks for checking this so quickly. Here's my feedback:

> wcadmin_task_view
> Perhaps we could add another event that only fires on loading the payment task?

That would be great. We can add a new event called: wcadmin_tasklist_payments_options

> wcadmin_tasklist_payment_setup
> and never includes woocommerce_payments since it always appears outside of the lists and in it's own card. We could amend this to include all visible gateways if preferred, but I haven't made this change yet.

It would be good to include WCPayments. But as long as it's included in the new event mentioned above we're good to go. As one of our goals with the tracking in the payments task is to answer the following questions: among all the users seeing X extension in the payments task, how many are installing it?

> wcadmin_tasklist_payment_install_method
> Firing as expected, but only when plugins are installed. This does not include BACS and COD since those don't require installation. This seems like the correct behavior semantically; do we want to change this?

I'm good with adding them to this event or create a new one just to track BACS and COD. As long as we can check how many users are enabling them. What do you think?

> wcadmin_tasklist_payment_connect_start
> This was being fired whenever a connect form was submitted

Can you provide more details about the current behavior? Are you saying that the event is only triggered when they submit a form on the Stripe/PayPay/ etc side? I'd like to have more clarity on this before evaluating if it makes sense to have two events. 

Thanks!

joelclimbsthings commented:
Thanks @pmcpinto . My thoughts below.

> wcadmin_task_view
Perhaps we could add another event that only fires on loading the payment task?
That would be great. We can add a new event called: wcadmin_tasklist_payments_options

Yup! Done, and it includes all visible gateways in the `options` property on the payload.

> wcadmin_tasklist_payment_setup
As one of our goals with the tracking in the payments task is to answer the following questions: among all the users seeing X extension in the payments task, how many are installing it?

Since I added the `wcadmin_tasklist_payments_options` event we likely don't need the list of IDs on this event anymore; is it safe to remove then, since it could just add confusion (since it doesn't show all of them)? 

Part of the confusion here is that there are a number of different ways for gateways to be configured/connected/enabled/etc. This particular event will only fire for those that require setup, and only with the "Setup" or "Finish Setup" buttons are clicked. The only event that I _believe_ will be fired for all gateways to indicate that they're enabled/connected is _wcadmin_tasklist_payment_connect_method_ as I understand it. Even for Slotfills, if they're implemented correctly, this will be fired when `markConfigured()` is executed. 

> wcadmin_tasklist_payment_connect_start

Before this PR this was being fired when a gateway was setup to display a form directly in the payments task (such as Stripe API keys), and the submit button was hit on that form. There is _another_ method that some gateways will connect that displays a "Connect" button that will then direct them to an outside site, such as with WCPay. I moved this to be fired when the "Connect" button was clicked instead, but I'm not really certain if that's what we want either at this point.

I'm going to amend the above description to include the exact conditions for each event when they fire, and update it as we go.

pmcpinto commented:
> Since I added the wcadmin_tasklist_payments_options event we likely don't need the list of IDs on this event anymore; is it safe to remove then, since it could just add confusion (since it doesn't show all of them)?

@joelclimbsthings Sounds good. Thanks for the explanation.

> I moved this to be fired when the "Connect" button was clicked instead, but I'm not really certain if that's what we want either at this point.

Yes, it makes more sense to fire it when they click the button, so we can measure their intent to proceed to the next step of the setup.

> I've noticed that some of the above events can occur more than once at a time. Does that matter when parsing the data?

No. AFAIK this won't be an issue.

joelclimbsthings commented on client/task-list/tasks/PaymentGatewaySuggestions/components/Setup/Configure.js:
```diff
@@ -129,7 +124,15 @@ export const Configure = ( {
+				<Button
+					isPrimary
+					onClick={ () =>
+						recordEvent( 'tasklist_payment_connect_start', {
```
Thanks, I'd wondered about that but had seen this pattern used elsewhere. We could alternatively add some kind of support for just delaying the link if an async function was provided for `onClick` as well. Likely better for an another PR.

joelclimbsthings commented:
Thanks for the early review @joshuatf . Looks like we're ready for a final review 👍🏻

mattsherman commented:
@joelclimbsthings @joshuatf Let's get this PR wrapped up and merged!

joelclimbsthings commented:
> Let's get this PR wrapped up and merged!

Yup @mattsherman . Been ready to go, just waiting on a review.

joshuatf commented on client/task-list/tasks/PaymentGatewaySuggestions/index.js:
```diff
@@ -82,6 +82,16 @@ export const PaymentGatewaySuggestions = ( { query } ) => {
 
+	useEffect( () => {
+		if ( paymentGateways.size ) {
+			recordEvent( 'tasklist_payments_options', {
```
This event fires on every single component update, which is quite a lot.  I think we'll need to memoize `mappedSuggestions` and possibly extract it out of `useSelect` to prevent this constant firing.

<img width="721" alt="Screen Shot 2021-07-15 at 10 28 40 AM" src="https://user-images.githubusercontent.com/10561050/125805904-222dbf7e-43b6-4ce8-8267-4a3049e9377b.png">

joelclimbsthings replied:
Good call @joshuatf , made this change in https://github.com/woocommerce/woocommerce-admin/pull/7304/commits/3181c7cec9dbb0399c4d0eb918f545b7510feff5.

joelclimbsthings commented:
Thanks for the review, @joshuatf . I made the requested change 👍🏻

joshuatf commented on client/task-list/tasks/PaymentGatewaySuggestions/index.js:
```diff
@@ -22,65 +22,76 @@ import './plugins/Bacs';
+	}, [] );
+
+	const paymentGateways = useMemo( () => fetchedPaymentGateways, [
+		fetchedPaymentGateways.size,
```
One of the issues with using `.size()` is that we won't get updates to the current list of enriched gateways.  We could also seem some edge cases in the future where the suggestions change, but are the same number of items and don't appear due to this memoization.

joshuatf reviewed (commented):
Hey @joelclimbsthings- thanks for the update!  Hope you don't mind, but I extracted your logic out of the `useSelect` in https://github.com/woocommerce/woocommerce-admin/pull/7304/commits/a4347ee8cc955e766bf7806523260492d6419ea0.  Reasons listed below.

Feel free to overwrite this or audit this as you please.

joelclimbsthings commented:
Great, appreciate the revision @joshuatf and glad to go with it. I like the simplified `select()` logic as well. I'll get this merged if you think it looks good now.

joshuatf reviewed (approved):
Thanks for the tracks fixes, Joel!  Testing well and LGTM :shipit:
